### PR TITLE
Add contiguous_strides as a correct replacement of defaultStride

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67833
* #68188
* #67996
* #68523
* #68522
* #69909
* #69908
* #68185
* #68184
* #68183
* #67014
* #66934
* __->__ #67789
* #66933

`at::defaultStride` was added in https://github.com/pytorch/pytorch/pull/18779.
As it was noted in that PR, it differs from the actual computation of
the default strides when one or more of the dimensions of the tensor are
zero. See https://github.com/pytorch/pytorch/pull/18779#discussion_r272296140

We add two functions, `contiguous_strides` and `contiguous_strides_vec`
which correct this issue and we replace the previous (wrong) uses of
`defaultStride`.

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

Differential Revision: [D32684852](https://our.internmc.facebook.com/intern/diff/D32684852)

### CODE REVIEW

**Comment by Lezcano** [comment]
@IvanYashchuk This correct the error we discussed for matrices with one of its sizes zero being zero. It also simplifies the writing of structured kernesl with fortran backends.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -222,15 +222,11 @@ TORCH_META_FUNC(triangular_solve)(const Tensor& self, const Tensor& A, bool uppe
     auto nrows = A.size(-2);
 
     // make column major strides for BLAS
-    auto solution_strides = at::detail::defaultStrides(self_broadcast_size);
-    solution_strides[ndim - 2] = 1;
-    solution_strides[ndim - 1] = nrows;
+    const auto solution_strides = at::native::contiguous_strides(self_broadcast_size, /*f-contig=*/true);
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:225]: Thanks for fixing that!
Why the tests were not failing? You mentioned that autograd doesn't like matrices with 0 strides and `sample_inputs_linalg_solve` from `common_methods_invocations.py` has 0x0 input matrix case.

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:225]: It's forward AD that was failing. Can it be the case that this function does not support forward AD?

**Review by mruberry — APPROVED** [review]
Cool!

Let's give @IvanYashchuk and @nikitaved a chance to review, too.

File: aten/src/ATen/native/LinearAlgebraUtils.h [hunk 2]
```diff
@@ -30,6 +32,46 @@ static char to_blas(TransposeType trans) {
   TORCH_INTERNAL_ASSERT(false, "Invalid transpose type");
 }
 
+template<class Vec>
+static inline Vec contiguous_strides_template(const IntArrayRef sizes, const bool f_contig=false) {
+  static_assert(std::is_same<IntArrayRef::value_type, typename Vec::value_type>::value,
+                "Incompatible integral type of sizes and strides");
+  // f_contig chooses between Fortran (F-contiguous) and C-contiguous strides
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/LinearAlgebraUtils.h:39]: Unfortunately, this comment is not correct.
`f_contig` chooses between C-contiguous strides (`f_contig=false`) and C-contiguous strides for N-2 dimensions + F-contiguous strides for the last 2 dimensions (`f_contig=true`).
This function is applicable only to batches of matrices and I think it should be reflected in the function name.

```python
In [1]: import numpy as np
# note that NumPy returns strides using bytes
In [2]: np.ones((2, 3, 4, 5, 6), order='C').strides
Out[2]: (2880, 960, 240, 48, 8)

In [3]: np.ones((2, 3, 4, 5, 6), order='F').strides
Out[3]: (8, 16, 48, 192, 960)
```

**Review by IvanYashchuk — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/LinearAlgebraUtils.h (hunk 2 above)
Comment (Lezcano) [reply aten/src/ATen/native/LinearAlgebraUtils.h:39]: That's a fair point. Corrected.

**Review by IvanYashchuk — APPROVED** [review]

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

File: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -2790,9 +2790,7 @@ std::tuple<Tensor, Tensor, Tensor> _svd_helper_cuda_legacy(const Tensor& self, b
   // _create_U_S_VT takes care of a part of these requirements (for U, S and VT)
   // For the input matrix, this requirements are being taken care of below.
   // Specify strides
-  auto self_col_major_strides = at::detail::defaultStrides(self.sizes());
-  self_col_major_strides[self.dim() - 2] = 1;
-  self_col_major_strides[self.dim() - 1] = m;
+  auto self_col_major_strides = at::native::contiguous_strides(self.sizes(), /*f_contig=*/true);
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2793]: Internal error for this change:

```
stderr: aten/src/ATen/native/cuda/BatchLinearAlgebra.cpp:2781:11: error: unused variable 'm' [-Werror,-Wunused-variable]
  int64_t m = self.size(-2);
          ^
```

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
```
stderr: aten/src/ATen/native/BatchLinearAlgebra.cpp:221:10: error: unused variable 'ndim' [-Werror,-Wunused-variable]
    auto ndim = self_broadcast_size.size();
         ^
aten/src/ATen/native/BatchLinearAlgebra.cpp:222:10: error: unused variable 'nrows' [-Werror,-Wunused-variable]
    auto nrows = A.size(-2);
         ^
stderr: aten/src/ATen/native/BatchLinearAlgebra.cpp:221:10: error: unused variable 'ndim' [-Werror,-Wunused-variable]
    auto ndim = self_broadcast_size.size();
         ^
aten/src/ATen/native/BatchLinearAlgebra.cpp:222:10: error: unused variable 'nrows' [-Werror,-Wunused-variable]
    auto nrows = A.size(-2);
```

Maybe we should build with these same flags, @dagitses @malfet, since they block landing?

**Comment by Lezcano** [comment]
I just fixed this one @mruberry. But yes, it'd be great to, at least, add this one flag to the open CI.

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Comment by mruberry** [comment]
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32684852).

**Pull request closed without merging**